coffee/rewriter.py

```diff
@@ -286,102 +412,92 @@ def __init__(self, stmt, expr_info, header, decls, hoisted, expr_graph):
         except ValueError:
             self._expr_handled.append(stmt)
             self.expr_id = self._expr_handled.index(stmt)
-        self.counter = 0
-
-    def _extract_exprs(self, node, expr_dep, length=0):
-        """Extract invariant sub-expressions from the original expression.
-        Hoistable sub-expressions are stored in expr_dep."""
 
-        def hoist(node, dep, expr_dep, _extract=True):
-            if _extract:
-                node = Par(node) if isinstance(node, Symbol) else node
-                expr_dep[dep].append(node)
-            self.extracted = self.extracted or _extract
+    def _extract(self, node, symbols, **kwargs):
+        """Extract invariant subexpressions from the original expression.
+        Hoistable sub-expressions are stored in ``dep_subexprs``."""
+
+        # Parameters driving the extraction pass
+        hoist_domain_const = kwargs.get('hoist_domain_const')
+        hoist_out_domain = kwargs.get('hoist_out_domain')
+
+        # Constants used to charaterize sub-expressions:
+        INVARIANT = 0  # expression is loop invariant
+        HOISTED = 1  # expression should not be hoisted any further
+
+        def __try_hoist(node, dep):
+            should_extract = True
+            if isinstance(node, Symbol):
+                should_extract = False
+            elif hoist_out_domain:
+                if not set(dep).issubset(set(self.expr_info.out_domain_dims)):
+                    should_extract = False
+            if should_extract:
+                dep_subexprs[dep].append(node)
+            return should_extract
+
+        def __extract(node, dep_subexprs):
```

Comment:
Yes, but I wrote it before Lawrence wrote the infrastructure, so maybe one day we can turn it into a visitor. Not gonna happen now.